enum with underlying type: char

``` r
cpp11::cpp_source(code = '
#include <cpp11.hpp>

enum class Color : char { red, green = 20, blue };

[[cpp11::register]]
void fun(SEXP value) {
  cpp11::as_cpp<Color>(value);
}

', quiet = FALSE)
#> clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I/Users/romainfrancois/.R/library/4.0/cpp11/include  -I/usr/local/include   -fPIC  -Wall -O3 -Wall -Wimplicit-int-float-conversion -c /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpzSpGdd/filef6b4538517d5/src/code_0.cpp -o /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpzSpGdd/filef6b4538517d5/src/code_0.o
#> In file included from /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpzSpGdd/filef6b4538517d5/src/code_0.cpp:2:
#> In file included from /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11.hpp:5:
#> /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/as.hpp:80:27: error: call to function 'as_cpp' that is neither visible in the template definition nor found by argument-dependent lookup
#>     return static_cast<E>(as_cpp<typename std::underlying_type<E>::type>(from));
#>                           ^
#> /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpzSpGdd/filef6b4538517d5/src/code_0.cpp:8:10: note: in instantiation of function template specialization 'cpp11::as_cpp<Color>' requested here
#>   cpp11::as_cpp<Color>(value);
#>          ^
#> /Users/romainfrancois/.R/library/4.0/cpp11/include/cpp11/as.hpp:143:12: note: 'as_cpp' should be declared prior to the call site or in the global namespace
#> is_char<T> as_cpp(SEXP from) {
#>            ^
#> 1 error generated.
#> make: *** [/private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/RtmpzSpGdd/filef6b4538517d5/src/code_0.o] Error 1
#> Error in dyn.load(shared_lib, local = TRUE, now = TRUE): unable to load shared object '/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T//RtmpzSpGdd/filef6b4538517d5/src/code_0.so':
#>   dlopen(/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T//RtmpzSpGdd/filef6b4538517d5/src/code_0.so, 6): image not found
fun(0L)
#> Error in .Call("_code_0_fun", value, PACKAGE = "code_0"): "_code_0_fun" not available for .Call() for package "code_0"
```

<sup>Created on 2020-08-05 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

--- a/cpp11test/src/test-as.cpp
+++ b/cpp11test/src/test-as.cpp
@@ -90,6 +90,19 @@ context("as_cpp-C++") {
     UNPROTECT(1);
   }
 
+  test_that("as_cpp<enum:char>(INTSEXP)") {
+    enum class Response: char { YES, NO, MAYBE };
+    SEXP r = PROTECT(Rf_allocVector(INTSXP, 1));
+
+    for (Response e : {Response::YES, Response::NO, Response::MAYBE, static_cast<Response>(42)}) {
+      INTEGER(r)[0] = static_cast<int>(e);
+      auto x = cpp11::as_cpp<Response>(r);
+      expect_true(x == e);
+    }
+
+    UNPROTECT(1);
+  }
+
   test_that("as_cpp<double>(REALSXP)") {
     SEXP r = PROTECT(Rf_allocVector(REALSXP, 1));
     REAL(r)[0] = 1.2;

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -96,7 +96,12 @@ enable_if_integral<T, T> as_cpp(SEXP from) {
 template <typename E>
 enable_if_enum<E, E> as_cpp(SEXP from) {
   if (Rf_isInteger(from)) {
-    return static_cast<E>(as_cpp<typename std::underlying_type<E>::type>(from));
+    using Type = typename std::conditional<
+      std::is_same<char, typename std::underlying_type<E>::type>::value,
+      int,
+      typename std::underlying_type<E>::type
+    >::type;
+    return static_cast<E>(as_cpp<Type>(from));
   }
 
   stop("Expected single integer value");

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -96,12 +96,13 @@ enable_if_integral<T, T> as_cpp(SEXP from) {
 template <typename E>
 enable_if_enum<E, E> as_cpp(SEXP from) {
   if (Rf_isInteger(from)) {
-    using Type = typename std::conditional<
-      std::is_same<char, typename std::underlying_type<E>::type>::value,
-      int,
-      typename std::underlying_type<E>::type
+    using underlying_type = typename std::underlying_type<E>::type;
+    using int_type = typename std::conditional<
+      std::is_same<char, underlying_type>::value,
+      int, // as_cpp<char> would trigger undesired string conversions
+      underlying_type
     >::type;
-    return static_cast<E>(as_cpp<Type>(from));
+    return static_cast<E>(as_cpp<int_type>(from));
   }
 
   stop("Expected single integer value");

